File: content/problems/0802-find-eventual-safe-states/solutions/dfs-color-cycle-propagation/solution.cpp
class Solution {
    bool safe(int node,vector<vector<int>>& graph,vector<int>& color){
        if(color[node])return color[node]==2;
        color[node]=1;
        for(int next:graph[node])if(!safe(next,graph,color)){
            color[node]=3;
            return false;
        }
        color[node]=2;
        return true;
    }
    public:vector<int> eventualSafeNodes(vector<vector<int>>& graph){
        vector<int> color(graph.size()),answer;
        for(int node=0;node<(int)graph.size();++node)if(safe(node,graph,color))answer.push_back(node);
        return answer;
    }
};

File: content/problems/0802-find-eventual-safe-states/solutions/dfs-color-cycle-propagation/Solution.kt
class Solution {
    fun eventualSafeNodes(graph:Array<IntArray>):List<Int>{
        val color=IntArray(graph.size);
        val answer=mutableListOf<Int>();
        for(node in graph.indices)if(safe(node,graph,color))answer.add(node);
        return answer
    };
    private fun safe(node:Int,graph:Array<IntArray>,color:IntArray):Boolean{
        if(color[node]!=0)return color[node]==2;
        color[node]=1;
        for(next in graph[node])if(!safe(next,graph,color)){
            color[node]=3;
            return false
        };
        color[node]=2;
        return true
    }
}
